File: Solutions/Random_Point_in_Non-overlapping_Rectangles/Random_Point_in_Non-overlapping_Rectangles.js
/**
 /**
 * @param {number[][]} rects
 */
var Solution = function(rects) {
    this.rects = rects;
    this.map = {};
    this.sum = 0;
    // we put in the map the number of points that belong to each rect
    for(let i in rects) {
        const rect = rects[i];
        // the number of points can be picked in this rectangle
        this.sum += (rect[2] - rect[0] + 1) * (rect[3] - rect[1] + 1);
        this.map[this.sum] = i;
    }
    this.keys = Object.keys(this.map);
};

/**
 * @return {number[]}
 */
Solution.prototype.pick = function() {
    // random point pick between [1, this.sum]
    const randomPointPick = Math.floor(Math.random() * this.sum) + 1;
    
    // we look for the randomPointPick in the keys of the map
    let pointInMap;
    // the keys exists in map
    if(this.map[randomPointPick]) pointInMap = randomPointPick;
    // the key is the first in the map (we do this check before doing binary search because its out of boundery)
    else if(randomPointPick < this.keys[0]) pointInMap = this.keys[0];
    let high = this.keys.length;
    let low = 1;
    // binary search to find the closest key that bigger than randomPointPick
    while(low <= high && !pointInMap) {
        const mid = Math.floor((low + (high-low)/2));
        if(randomPointPick > this.keys[mid-1] && randomPointPick < this.keys[mid]) {
            pointInMap = this.keys[mid];
            break;
        } else if (randomPointPick > this.keys[mid]){
            low = mid+1;
        } else {
            high = mid-1;
        }
    }
    
    // we have the point, now we can get which rect belong to that point
    const pointInRects = this.map[pointInMap];
    const chosen = this.rects[pointInRects];
    const rightX = chosen[2];
    const leftX = chosen[0];
    const topY = chosen[3];
    const bottomY = chosen[1];
    const pickX = Math.floor(Math.random() * (rightX-leftX+1)) + leftX;
    const pickY = Math.floor(Math.random() * (topY-bottomY+1)) + bottomY;
    return [pickX, pickY]
};


//Another. Image in images: Example-of-alias-method_js_second_solution.png
//Original post: https://leetcode.com/problems/random-point-in-non-overlapping-rectangles/discuss/805452/Javascript-O(n)-initiation-O(1)-picking-using-Alias-Method
/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(rects)
 * var param_1 = obj.pick()
 */

var Solution = function(rects) {
    // Imagine we put the boxes into a 2D pool and pick one. Then the bigger box will more likely to be seen and picked
    // So first we initialize the pool with the area of the boxes
    let totalArea = 0;
    this.boxes = rects.map(([x1, y1, x2, y2], index, area) => (
      totalArea += area = (x2 - x1 + 1) * (y2 - y1 + 1),
      { area, index, x1, y1, x2, y2 }
    ));
    // Alias method says that, for example you have n boxes, and n rooms with can contain exactly the average area of the boxes,
    // you can distribute the content of the boxes into the rooms such that each room contains only content of 2 type of boxes.
    // So the idea is putting each box into each room. Then some box will be larger than the room, and some will be smaller
    // You can then split the larger box and put into the room which contains the smaller box
    // So now let's classify the boxes base on if it's larger or smaller than the room size
    const avgArea = this.avg = totalArea / rects.length;
    const smallers = this.boxes.filter(box => box.area < avgArea);
    const largers = this.boxes.filter(box => box.area > avgArea);
    while (smallers.length !== 0 && largers.length !== 0) {
      const smaller = smallers.shift();
      const larger = largers[0];
      smaller.added = larger.index;
      larger.area -= avgArea - smaller.area;
      // After splitting the larger box, calling A for example, and put into the room of the smaller box, that room is now full, and we no longer care about it
      // About the box A, if it is now still bigger than the room capacity, keep it in the "largers" stack to work with later
      // If it is now fit into its room, forget about it
      // If it becomes smaller for a typical room, we put it into the "smallers" stack to find it a new "larger boxie"
      larger.area <= avgArea && (
        largers.shift(),
        larger.area < avgArea && (smallers[smallers.length] = larger)
      );
    }
    this.pick = function() {
      // Now we just need to pick a random room
      const box = this.boxes[Math.random() * this.boxes.length | 0];
      // And then pick 1 of the 2 boxes in that room, based on their area as above
      const { x1, y1, x2, y2 } = Math.random() * this.avg > box.area ? this.boxes[box.added] : box;
      // Finally return a point inside that box
      return [
        Math.floor(x1 + Math.random() * (x2 - x1 + 1)),
        Math.floor(y1 + Math.random() * (y2 - y1 + 1))
      ];
    }
  };

File: Solutions/Random_Point_in_Non-overlapping_Rectangles/Random_Point_in_Non-overlapping_Rectangles.cpp
class Solution {
public:
    vector<int> np;
    vector<vector<int>> Rects;
    Solution(vector<vector<int>>& rects) {
        Rects = rects;
        for(auto rect : rects){
            int l1 = rect[2] - rect[0] + 1;
            int l2 = rect[3] - rect[1] + 1;
            int val = np.size() ? np.back() + (l1*l2) : l1*l2; 
            np.push_back(val);
        }
    }
    
    vector<int> pick() {
        int m = np.back();
        int r = rand() % m;
        auto it = upper_bound(np.begin(), np.end(), r);
        int rect = it - np.begin();  //end of step 1
		//step 2 begins
        vector<int> R = Rects[rect];
        int x = rand() % (R[2]-R[0]+1) + R[0];
        int y = rand() % (R[3]-R[1]+1) + R[1];
        return {x, y};
    }
};

//Fastest
class Solution {
    unsigned int areas[100];
    unsigned int i = 0;
    default_random_engine gen;
    uniform_int_distribution<> distrib;
    vector<vector<int>> rs;
public:
    Solution(vector<vector<int>>& rects) {
        unsigned int tot = 0;
        for (const auto &r: rects) 
            areas[i++] = tot += (unsigned int)(1+r[2]-r[0])*(unsigned int)(1+r[3]-r[1]);
        distrib.param(uniform_int_distribution<>::param_type(0,tot-1));
        rs = move(rects);
    }
    
    vector<int> pick() {
        unsigned int r = distrib(gen);
        // cout << "r: " << r << endl;
        const auto ra = upper_bound(&areas[0], &areas[0]+i, r);
        r = *ra - r - 1;
        // cout << "ra: " << *ra << " new r: " << r << endl;
        const auto &tr = rs[ra - &areas[0]];
        const unsigned int w = 1+tr[2]-tr[0];
        const unsigned int y = r/w;
        const unsigned int x = r%w;
        // cout << "w: " << w << " x: " << x << " y: " << y << endl;
        return {static_cast<int>(x+tr[0]),static_cast<int>(y+tr[1])};
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution* obj = new Solution(rects);
 * vector<int> param_1 = obj->pick();
 */

static int faster=[](){std::ios::sync_with_stdio(false);std::cin.tie(nullptr);return 13;}();


//Another good
class Solution {
    vector<int> csum; //cumulative sum of points of rectangles
    vector<vector<int>> _rects; 
public:
    Solution(vector<vector<int>>& rects) { 
        // calculuate cumulative sum: 
        for (auto r: rects)
        {
            int points = ( r[2]-r[0]+1)* (r[3]-r[1]+1); 
            if (csum.size()==0)
                csum.push_back(points);
            else
                csum.push_back(points + csum.back());
        }
        _rects=rects; 
    }
    
    // 5 mistakes!!
    vector<int> pick() {
        // randomly pick one rectangle based on weights: points        
        int p = rand() %csum.back();
        
        // convert it into the points in a rectangle
        // need to calculate the distance of iterators !!
        int rect_id = upper_bound (csum.begin(), csum.end(), p) - csum.begin(); 
        
        // offset within the rectangle 
        int inner_id ; 
        if (rect_id==0) inner_id = p;
        else
            inner_id= p - csum[rect_id - 1];
        
        // convert it into a 2-D coordinates
        int v0= _rects[rect_id][0], v1=_rects[rect_id][1];
        int v2= _rects[rect_id][2], v3=_rects[rect_id][3];
        
        int col= v3-v1+1; 
        int i = inner_id/col + v0;         
        int j = inner_id%col + v1; 
        
        return {i,j}; 
    }
};